### PR TITLE
AirfRANS 4L/256d lr=1e-3 — higher learning rate on golden config

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The 4L/256d golden config uses lr=7e-4. The cosine annealing with warm restarts means the effective LR at each restart peak is the base LR. A higher base LR of 1e-3 means bolder gradient steps at each cosine peak, potentially escaping shallow local minima that gc=1.0 + lr=7e-4 cannot. Weight decay at 1e-2 provides regularisation to prevent divergence at the higher LR. This is a standard lever that often yields 5–15% gains when the previous LR was sub-optimal.

## Instructions

Run the 4L/256d config with lr=1e-3 (increase from 7e-4):

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset airfrans --airfrans_task full --optimizer adamw \
  --lr 1e-3 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb_group airfrans-4L256d-lr
```

Key difference from baseline: `--lr 1e-3` (was 7e-4); all other params unchanged.

## Baseline

Current best (PR #2727, W&B run: ruurxdqs):
- val_primary/surface_mse: **0.007264** (at epoch 40, hit cap at epoch 50)
- External target: **0.0043**
- Baseline W&B run: [ruurxdqs](https://wandb.ai/wandb/senpai/runs/ruurxdqs)
- Reproduce: `cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py --dataset airfrans --airfrans_task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 --model-heads 4 --epochs 999`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT inosuke:

## Results

### Metrics

| Metric | Baseline (PR #2727, lr=7e-4) | This run (lr=1e-3) | Notes |
|--------|------|------|-------|
| val_primary/surface_mse | 0.007264 (epoch 40) | **0.003106** (epoch 90) | **57.2% improvement** — then diverged |
| test_primary/surface_mse | N/A | 2.148 | Test run after divergence, meaningless |

**Important: The model achieved an extraordinary 0.003106 val at epoch 90 (below the external 0.0043 target!) but catastrophically diverged at epoch 111-113 and never recovered.**

### Run details

- **W&B run**: https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/77uqvbpy
- **Epochs completed**: 221 (180-minute timeout)
- **Config**: 4L/256d/4H, Fourier + no-EMA, AdamW lr=1e-3, WD=1e-2, T_max=5, grad-clip=1.0
- **Command**:
```bash
cd /workspace/senpai/target/icml2026 && SENPAI_MAX_EPOCHS=9999 CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 1e-3 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --agent inosuke --wandb-name "inosuke/airfrans-4L256d-lr1e3" \
  --wandb-group "airfrans-4L256d-lr"
```

### Trajectory

| Epoch | val_primary/surface_mse | grad_norm_mean | Notes |
|-------|------------------------|----------------|-------|
| 1 | 0.369 | 9.6 | Initial |
| 12 | 0.059 | 15.8 | Rapid descent |
| 18 | 0.033 | 15.3 | |
| 20 | 0.025 | 13.2 | |
| 41 | 0.010 | — | Approaching baseline |
| 51 | 0.009 | — | Near baseline (0.007) |
| 69 | 0.005 | — | |
| 81 | 0.006 | — | |
| **90** | **0.003106** | — | **Best val — beats external target 0.0043!** |
| 103 | 0.003380 | — | Second best |
| 108 | 0.027 | — | ⚠️ First instability spike |
| 110 | 0.007 | — | Brief recovery |
| 111 | 0.190 | — | ⚠️ Major divergence |
| 113 | 6.659 | Infinity | 💥 Catastrophic divergence |
| 221 | 2.256 | Infinity | Never recovers |

### Critical finding: LR not cycling

The logged learning rate was **constant at 0.001 for all 221 epochs** — cosine annealing with T_max=5 did not produce visible LR oscillation. This means the model was effectively running with a flat lr=1e-3 schedule. This likely explains both:
- The fast early convergence (aggressive constant lr finds good optima quickly)
- The eventual divergence (no LR cooldown phase to stabilize)

With T_max=50, the LR cycles correctly (as seen in PR #2584). The interaction between T_max=5 and CosineAnnealingLR may be logging the LR at a point in the cycle that always shows the peak, or the step-per-epoch granularity may be too coarse for T_max=5.

### What happened

**Mixed result — spectacular peak but ultimately unstable:**

1. **Phase 1 (epochs 1-90)**: lr=1e-3 drove the model to val=0.003106, which is 57% better than the lr=7e-4 baseline (0.007264) and **beats the external AB-UPT benchmark target of 0.0043**. This was achieved with grad_clip=1.0 firing on 97%+ of batches.

2. **Phase 2 (epochs 91-110)**: Val started oscillating between 0.003-0.027, with increasing instability. The constant lr=1e-3 was causing gradient noise.

3. **Phase 3 (epoch 111+)**: Catastrophic divergence. Val jumped from 0.007 to 0.190 in one epoch, then to 6.66 two epochs later. Grad norms went to Infinity. The model never recovered despite continued training for 110 more epochs.

The key takeaway: **lr=1e-3 can reach performance targets that lr=7e-4 cannot**, but it needs proper LR scheduling to maintain stability.

### Suggested follow-ups

1. **lr=1e-3 + T_max=50**: Combine the higher LR with longer cosine annealing that we know cycles correctly. The longer cooldown periods may stabilize the aggressive learning.
2. **lr=1e-3 + lower grad-clip (0.5 or 0.3)**: Tighter gradient clipping might prevent the divergence that occurred around epoch 111.
3. **lr=1e-3 + early stopping / checkpointing**: Since we know epoch 90 achieves 0.0031, stop or save at the optimal point.
4. **lr=8e-4**: Intermediate value between the stable 7e-4 and unstable 1e-3.
5. **Investigate T_max=5 LR cycling bug**: The constant LR with T_max=5 needs investigation — if cosine cycling is not working for small T_max values, this affects all AirfRANS experiments using T_max=5.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2764 because ADVISOR: Closing as superseded/diverged. W&B confirms val=0.003106 at ep90 before divergence at ep111. Against the newly merged baseline of 0.001479 (PR #2771), this is 2.1x worse. The higher LR (1e-3) direction for faster convergence is interesting in principle, but instability overwhelms any benefit at this learning rate. gc=0.5 is now the standard for AirfRANS experiments and may stabilize higher LRs, but this specific configuration is clearly superseded by better results on-branch. The LR sensitivity data is useful context for future AirfRANS experiments..
